OpenShift: Fix `quarkus.openshift.arguments` property

Sgitario opened the pull request:
At the moment, when using the `quarkus.openshift.arguments` property and an unknown base image, it's replacing the arguments that are required for the Java command and hence making the application fail. For example, with no arguments are supplied, the OpenShift.yaml looks like:

```
command: java
args:
    - '-Dquarkus.http.host=0.0.0.0'
    - '-Djava.util.logging.manager=org.jboss.logmanager.LogManager'
    - '-jar'
    - /deployments/quarkus-run.jar
```

If we now set `quarkus.openshift.arguments=A,B`, then it changes to:

```
command: java
args:
    - 'A'
    - 'B'
```

And then, the above configuration would make our deployment fail (because the jar is missing).

This PR addresses the above situation by moving the required JAVA arguments to be part of the command and leaving the args field empty:

```
command: 
    - java
    - '-Dquarkus.http.host=0.0.0.0'
    - '-Djava.util.logging.manager=org.jboss.logmanager.LogManager'
    - '-jar'
    - /deployments/quarkus-run.jar
```

And now if we supply the arguments: `quarkus.openshift.arguments=A,B`, then:

```
command: 
    - java
    - '-Dquarkus.http.host=0.0.0.0'
    - '-Djava.util.logging.manager=org.jboss.logmanager.LogManager'
    - '-jar'
    - /deployments/quarkus-run.jar
args:
    - 'A'
    - 'B'
```

Users would still be able to replace the whole command by using the command property: `quarkus.openshift.command` which I think it's the expected behaviour. Now the command and argument properties work the same in Kubernetes and OpenShift deployments, regardless the base image we use. 

I've covered these changes using integration tests with the following scenarios:
- Kubernetes when using arguments
- Kubernetes when using command
- OpenShift when using arguments
- OpenShift when using default command
- OpenShift when using commands

Also, I've verified these changes manually by deploying a sample application (Rest ++ Command Line application) and now it works fine when supplying or not supplying custom arguments.

Fix https://github.com/quarkusio/quarkus/issues/15943

Sgitario commented:
cc @iocanel @geoand

iocanel commented:
> Users would still be able to replace the whole command by using the command property: quarkus.openshift.command which I think it's the expected behaviour.

@Sgitario: Can you please elaborate on this? I always considered the use of `command` and `arguments` to be quite literal (e.g. command: the path to the binary. arguments: the actual app arguments). Of course, this might be just my perception and I may totally miss the intended use. @geoand wdyt?

Sgitario commented:
> > Users would still be able to replace the whole command by using the command property: quarkus.openshift.command which I think it's the expected behaviour.
> 
> @Sgitario: Can you please elaborate on this? I always considered the use of `command` and `arguments` to be quite literal (e.g. command: the path to the binary. arguments: the actual app arguments). Of course, this might be just my perception and I may totally miss the intended use. @geoand wdyt?

As far I see it, the problem was that we were mixing up the command with arguments (the command was "java" and the arguments "-jar quarkus-run.jar") which from my point of view does not make totally sense.

With these changes, the command is "java -jar quarkus-run.jar" and the arguments are empty by default. Therefore, now they are indeed literal. Users are still able to overwrite the command and arguments, but now they are not tied up with each other.

iocanel commented:
> As far I see it, the problem was that we were mixing up the command with arguments (the command was "java" and the arguments "-jar quarkus-run.jar") which from my point of view does not make totally sense.

From my point of view, this makes total sense.
However, I see that command is not a `String` but a `List<String>` which strongly implies that `command` is not just the binary.
So, I would like a second opinion before I change my mind on this. @geoand @Ladicek wdyt?

geoand commented:
I would like use to do what we do in `JibConfig`m where we have `List<String> jvmArguments` that are **additive** to whatever is done by other parts of the image generation process, but we also have `Optional<List<String>> jvmEntrypoint` which allows a user to specify whatever launch command they like (so it could be a bash script with arguments)

iocanel commented:
> I would like use to do what we do in `JibConfig`m where we have `List<String> jvmArguments` that are **additive** to whatever is done by other parts of the image generation process, but we also have `Optional<List<String>> jvmEntrypoint` which allows a user to specify whatever launch command they like (so it could be a bash script with arguments)

Wouldn't that cause confusion between jvm and native modes?

geoand commented:
> > I would like use to do what we do in `JibConfig`m where we have `List<String> jvmArguments` that are **additive** to whatever is done by other parts of the image generation process, but we also have `Optional<List<String>> jvmEntrypoint` which allows a user to specify whatever launch command they like (so it could be a bash script with arguments)
> 
> Wouldn't that cause confusion between jvm and native modes?

Which is why Jib has separate options for jvm and native :)

iocanel commented:
> > > I would like use to do what we do in `JibConfig`m where we have `List<String> jvmArguments` that are **additive** to whatever is done by other parts of the image generation process, but we also have `Optional<List<String>> jvmEntrypoint` which allows a user to specify whatever launch command they like (so it could be a bash script with arguments)
> > 
> > 
> > Wouldn't that cause confusion between jvm and native modes?
> 
> Which is why Jib has separate options for jvm and native :)

This is however, not the case for our configuration files and I am afraid that things will get too complicated.

Sgitario commented:
> > > > I would like use to do what we do in `JibConfig`m where we have `List<String> jvmArguments` that are **additive** to whatever is done by other parts of the image generation process, but we also have `Optional<List<String>> jvmEntrypoint` which allows a user to specify whatever launch command they like (so it could be a bash script with arguments)
> > > 
> > > 
> > > Wouldn't that cause confusion between jvm and native modes?
> > 
> > 
> > Which is why Jib has separate options for jvm and native :)
> 
> This is however, not the case for our configuration files and I am afraid that things will get too complicated.

At first, I thought to add a new property `quarkus.openshift.app-arguments` or similar, but then my concern was that this was getting indeed too complicated imho taking into account that there is already an existing `quarkus.openshift.arguments` that only works when we use known base images.

geoand commented:
> This is however, not the case for our configuration files and I am afraid that things will get too complicated.

In that case, I agree with your proposal :)

Ladicek commented:
I think the proposal here kinda sorta makes sense, because according to https://kubernetes.io/docs/tasks/inject-data-application/define-command-argument-container/#notes, `command` is akin to image's `ENTRYPOINT`, while `args` are akin to image's `CMD`. But I didn't really think this through, there may easily be huge gaps in my understanding :-)

iocanel commented:
Thanks @Ladicek that really helps.

According to the link provided, the command pretty much corresponds to the Dockerfile entrypoint and arguments to Dockerfile cmd. Some additional clarifications are also added:

```
When you override the default Entrypoint and Cmd, these rules apply:
If you do not supply command or args for a Container, the defaults defined in the Docker image are used.
If you supply a command but no args for a Container, only the supplied command is used. The default EntryPoint and the default Cmd defined in the Docker image are ignored.
If you supply only args for a Container, the default Entrypoint defined in the Docker image is run with the args that you supplied.
If you supply a command and args, the default Entrypoint and the default Cmd defined in the Docker image are ignored. Your command is run with your args.
```

Now how cmd and entrypoint are meant to be used in dockerland is another story, but I feel that `entrypoint` corresponds to  the `what` the image is about and is something that isn't meant to be changed per invocation, while `cmd` corresponds to the `how` (e.g. application args & parameters?).

This means that I am starting to accept what @Sgitario suggests and the only real question if this is a very intrusive change is going to break things for users.

Sgitario commented:
> Thanks @Ladicek that really helps.
> 
> According to the link provided, the command pretty much corresponds to the Dockerfile entrypoint and arguments to Dockerfile cmd. Some additional clarifications are also added:
> 
> ```
> When you override the default Entrypoint and Cmd, these rules apply:
> If you do not supply command or args for a Container, the defaults defined in the Docker image are used.
> If you supply a command but no args for a Container, only the supplied command is used. The default EntryPoint and the default Cmd defined in the Docker image are ignored.
> If you supply only args for a Container, the default Entrypoint defined in the Docker image is run with the args that you supplied.
> If you supply a command and args, the default Entrypoint and the default Cmd defined in the Docker image are ignored. Your command is run with your args.
> ```
> 
> Now how cmd and entrypoint are meant to be used in dockerland is another story, but I feel that `entrypoint` corresponds to the `what` the image is about and is something that isn't meant to be changed per invocation, while `cmd` corresponds to the `how` (e.g. application args & parameters?).
> 
> This means that I am starting to accept what @Sgitario suggests and the only real question if this is a very intrusive change is going to break things for users.

This could be a breaking change for users that were using the OpenShift extension, an unknown base image and the `quarkus.openshift.arguments` property to tweak the Java command:

```
command: java
args:
    - '-Dquarkus.http.host=0.0.0.0'
    - '-DmyUserChange=anotherValue`
    - '-Djava.util.logging.manager=org.jboss.logmanager.LogManager'
    - '-jar'
    - /deployments/quarkus-run.jar
    - ARG1
```

With these changes, they should overwrite the Java command using the `quarkus.openshift.command` (and include the binary). Being said that, I don't think users are doing so or should have been doing so.

There are a failing test that I'm trying to investigate, but I can't reproduce locally :/

iocanel commented on extensions/kubernetes/vanilla/deployment/src/main/java/io/quarkus/kubernetes/deployment/KubernetesCommonHelper.java:
```diff
@@ -189,6 +188,50 @@
+    private static List<DecoratorBuildItem> createArgsDecorator(Optional<Project> project, String target, String name,
+            PlatformConfiguration config, Optional<KubernetesCommandBuildItem> command) {
+        List<String> args = new ArrayList<>();
+        command.ifPresent(cmd -> args.addAll(Arrays.asList(cmd.getArgs())));
```
Topic for discussion: Since, we are slightly changing the way we perceive `arguments` do we expect extension to contribute to the them? It kind of feels that now extensions are expected to only toy with `command`. Thoughts?

Sgitario replied:
For native, extensions still need to provide both the command and the arguments.
But I like what you say which would imply split KubernetesCommandBuildItem in KubernetesCommandBuildItem (only for the command) and KubernetesArgsBuildItem (only for the args).
Are you ok with and do you want me to do it as part of this PR?

geoand replied:
Why do we need to change `KubernetesCommandBuildItem`? I didn't get that part

iocanel replied:
We don't need to change anything.
I am just wondering / thinking out loud.

Do we expect extensions/build items to ever contribute to arguments?

Sgitario replied:
Yes, for Native we need extensions to contribute to the arguments.

geoand replied:
I don't see why the BuildItem now uses arrrays for both the command and args... Seems counterintuitive...

Sgitario replied:
I agree with you, I would change it to only have a single constructor for the command: `KubernetesCommandBuildItem(String ... cmd)` and add a method to set the arguments (as this is now optional for extensions).
Wdyt?

geoand replied:
Makes sense to me

Sgitario replied:
I've already refactored the `KubernetesCommandBuildItem`.

geoand commented:
When done, please squash the commits

Sgitario commented:
@iocanel I could find the root cause of the failing tests: the problem was caused by registering an args decorator with no args. This is obviously wrong and I've already fixed it.

Would you mind to take another look?

geoand commented:
Do we really have to use the Maven invoker style of tests?

I know you guys where looking into some issue with the QuarkusProdMode test, what was the result of that investigation? Can the PR be updated to move the tests to that style?

Sgitario commented:
> Do we really have to use the Maven invoker style of tests?
> 
> I know you guys where looking into some issue with the QuarkusProdMode test, what was the result of that investigation? Can the PR be updated to move the tests to that style?

The issue we were facing was unrelated to QuarkusProdMode and has been already fixed.
Afaik, I don't have any objections to move the scenarios from maven invoker to standard way.
Do you want me to change this?

geoand commented:
Good to hear.

Yes please move them. I initially developed QuarkusProdModeTest out of frustration with the Maven invoker tests, so I would really like to avoid going back to them 😎

Sgitario commented:
> Good to hear.
> 
> Yes please move them. I initially developed QuarkusProdModeTest out of frustration with the Maven invoker tests, so I would really like to avoid going back to them

Moved! I think it looks better and simpler now.

geoand commented:
Great, thanks 👌

geoand reviewed (approved):
LGTM, but I want @iocanel to have the final say

geoand commented:
There seems to be a memory leak in `QuarkusProdModeTest`... I'll have to investigate.

Sorry for the inconvinience!

geoand commented:
https://github.com/quarkusio/quarkus/pull/17956 fixes the issue. Once that is in, please rebase this PR onto `main`

geoand commented:
All set, can you please rebase?

Sgitario commented:
> All set, can you please rebase?

Rebased

gastaldi commented on extensions/container-image/container-image-openshift/deployment/src/main/java/io/quarkus/container/image/openshift/deployment/OpenshiftProcessor.java:
```diff
@@ -148,11 +148,12 @@ public void openshiftRequirementsJvm(OpenshiftConfig openshiftConfig,
+                cmd.addAll(Arrays.asList("-jar", pathToJar));
                 envProducer.produce(KubernetesEnvBuildItem.createSimpleVar("JAVA_APP_JAR", pathToJar, null));
-                commandProducer.produce(new KubernetesCommandBuildItem("java", args.toArray(new String[args.size()])));
+                commandProducer.produce(new KubernetesCommandBuildItem(cmd.toArray(new String[cmd.size()])));
```
```suggestion
                commandProducer.produce(new KubernetesCommandBuildItem(cmd));
```

Sgitario replied:
KubernetesCommandBuildItem requires an Array, not a List.

gastaldi replied:
If the command is converted back to a list internally, why not take a List instead?

Sgitario replied:
With array, we have the ability to write commands like ``KubernetesCommandBuildItem(new String[] { "java", "-jar", "jar.jar" })`("java")` or `KubernetesCommandBuildItem("java", "-jar", "jar.jar")` or `KubernetesCommandBuildItem(new String[] { "java", "-jar", "jar.jar" })`. This is how it was originally designed and didn't want to change as it would require to change other parts of the code as well.

We could certainly add a new constructor to accept a list and internally keeping the list instead of an array, but as I said, it would require to change the consumers of the KubernetesCommandBuildItem.

gastaldi replied:
I think it all depends on how it is used right now, really. The existing code seems to use a list for the arguments, so using an array might seem unnecessary here.

Sgitario replied:
In other parts, we're using the arrays: https://github.com/quarkusio/quarkus/pull/17889/files#diff-c15bcfcc919c2e97bb2cb66a3a836eb0537fbfb3a3fcfbd3373ad5df09749b63R206

Sgitario replied:
I've introduced a few factory methods in KubernetesCommandBuildItem to use either a command of array, or a list, or a combination of a command and arguments.
Let me know how looks like now.

maxandersen commented:
if i understand it right this is a breaking change (for good reasons) so should have entry in migration guide, right?

geoand commented:
Exactly.

@Sgitario can you write something here that we will put in the migration guide?

Sgitario commented:
> Exactly.
> 
> @Sgitario can you write something here that we will put in the migration guide?

I was figuring out how to add it myself. This is the content that I wrote:

```
### Java command and arguments mapping

When deploying a Quarkus application into OpenShift, by default the container configuration to run the application looked like:

```
command: java
args:
    - '-Dquarkus.http.host=0.0.0.0'
    - '-Djava.util.logging.manager=org.jboss.logmanager.LogManager'
    - '-jar'
    - /deployments/quarkus-run.jar
```

Both `command` and `args` fields can be overriden via the `quarkus.openshift.command` and `quarkus.openshift.arguments` properties. 

However, users can't simple append their custom Java arguments, for example: `java -jar /deployments/quarkus-run.jar param1 param2`. The only workaround was to copy and paste the existing content in the default `args` field and append the custom arguments `param1` and `param2`.

This has been fixed in 2.0 by moving all the Java related parameters into the `command` field:

```
command: 
    - 'java'
    - '-Dquarkus.http.host=0.0.0.0'
    - '-Djava.util.logging.manager=org.jboss.logmanager.LogManager'
    - '-jar'
    - /deployments/quarkus-run.jar
```

Now, users can use the `quarkus.openshift.arguments` property to append their custom Java arguments, for example: `quarkus.openshift.arguments=param1,param2`.
```

wdyt?

gastaldi commented on extensions/kubernetes/spi/src/main/java/io/quarkus/kubernetes/spi/KubernetesCommandBuildItem.java:
```diff
@@ -1,22 +1,30 @@
-    private final String command;
-    private final String[] args;
+    private final List<String> command;
+    private List<String> args = Collections.emptyList();
```
Please don't make it non-final. Instead use the constructor, static factory methods or introduce a builder

Sgitario replied:
Changed. 
I've created a few factory methods.
Let me know what you think now.

Sgitario committed:
OpenShift: Fix `quarkus.openshift.arguments` property

At the moment, when using the `quarkus.openshift.arguments` property and an unknown base image, it's replacing some arguments that are required for the Java command and hence making the application fail. For example, with no arguments are supplied, the OpenShift.yaml looks like:

```
command: java
args:
    - '-Dquarkus.http.host=0.0.0.0'
    - '-Djava.util.logging.manager=org.jboss.logmanager.LogManager'
    - '-jar'
    - /deployments/quarkus-run.jar
```

If we now set `quarkus.openshift.arguments=A,B`, then it changes to:

```
command: java
args:
    - 'A'
    - 'B'
```

This PR addresses the above situation by moving the required JAVA arguments to be part of the command and leaving the args field empty:

```
command: 
    - java
    - '-Dquarkus.http.host=0.0.0.0'
    - '-Djava.util.logging.manager=org.jboss.logmanager.LogManager'
    - '-jar'
    - /deployments/quarkus-run.jar
```

And now if we supply the arguments: `quarkus.openshift.arguments=A,B`, then:

```
command: 
    - java -Dquarkus.http.host=0.0.0.0 -Djava.util.logging.manager=org.jboss.logmanager.LogManager -jar /deployments/quarkus-run.jar
args:
    - 'A'
    - 'B'
```

Users would still be able to replace the whole command by using the command property: `quarkus.openshift.command` which I think it's the expected behaviour. Now the command and argument properties work the same in Kubernetes and OpenShift, regardless the base image we use for s2i. 

I've covered these changes using integration tests with the following scenarios:
- Kubernetes when using arguments
- Kubernetes when using command
- OpenShift when using arguments
- OpenShift when using commands

Also, I've verified these changes manually by deploying a sample application (Rest ++ Command Line application) and now it works fine.

gastaldi reviewed (approved):
LGTM

gsmet commented:
@iocanel can you check this and give us your blessing?

@geoand looks like something we would want for 2.0 or not?

geoand commented:
> @geoand looks like something we would want for 2.0 or not?

We could, although I don't think anyone will fall in despair if we don't :)

gsmet commented:
@iocanel ping?

geoand commented:
Where are we on this one?

gsmet commented:
Merging after discussion with @geoand . Thanks @Sgitario .
